theano/gpuarray/dnn.py

```diff
@@ -3132,12 +3133,13 @@ def local_abstractconv_cudnn(node):
     ctx = infer_context_name(*node.inputs)
     if not isinstance(node.inputs[0].type, GpuArrayType):
         return
-    if node.op.unshared:
-        return None
-    if isinstance(node.op, AbstractConv2d):
-        return local_abstractconv_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
-    elif isinstance(node.op, AbstractConv3d):
-        return local_abstractconv3d_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
+    with inherit_stack_trace(node.outputs):
+        if node.op.unshared:
+            return None
```

Comment:
Ideally moce this if outside the with statement.

Reply:
That's true. Sure.

Reply:
This is actually a `if .. else` block:
```python
with inherit_stack_trace(node.outputs):
        if node.op.unshared:
            return None
        if isinstance(node.op, AbstractConv2d):
            return local_abstractconv_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
        elif isinstance(node.op, AbstractConv3d):
            return local_abstractconv3d_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
```

So if I want to move the if outside the with, I will have to do something like this:

```python
if node.op.unshared:
	with inherit_stack_trace(node.outputs):
                return None
if isinstance(node.op, AbstractConv2d):
	with inherit_stack_trace(node.outputs):
               return local_abstractconv_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
elif isinstance(node.op, AbstractConv3d):
	with inherit_stack_trace(node.outputs):
                return local_abstractconv3d_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
```

Reply:
What about
```
if node.op.unshared:
        return None
if isinstance(node.op, AbstractConv2d):
	with inherit_stack_trace(node.outputs):
               return local_abstractconv_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
elif isinstance(node.op, AbstractConv3d):
	with inherit_stack_trace(node.outputs):
                return local_abstractconv3d_cudnn_graph(node.op, ctx, node.inputs, node.outputs)
```

As this is an optimization, when it return none, it mean it do nothing. So no new node build, do no new stack trace needed.

Reply:
I see! Thanks.